### PR TITLE
swarm, main, api: split sync flag into push and pull flags

### LINKED ISSUE #2196: add push-sync flag
The tests run during [phase 3 of Global Pinning](https://github.com/ethersphere/swarm/issues/2173) require pull and push sync to be independently set, but currently the `no-sync` flag sets both to the same value.

This has been manually patched in the testing branch, but needs to be properly implemented.

This can be merged into `master` and later into `global-pinning`.

### PR DESCRIPTION
**Pull request by mortelli**
This PR separates the `SwarmNoSyncFlag` into 2 independent flags: `SwarmNoPullSyncFlag` and `SwarmNoPushSyncFlag`.

closes #2196

### CODE REVIEW

File: swarm.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func NewSwarm(config *api.Config, mockStore *mock.NodeStore) (self *Swarm, err e
 		HiveParams:   config.HiveParams,
 		LightNode:    config.LightNodeEnabled,
 		BootnodeMode: config.BootnodeMode,
-		SyncEnabled:  config.SyncEnabled,
+		SyncEnabled:  config.PullSyncEnabled, // should config.PushSyncEnabled have an effect here as well?
```
Comment (mortelli) [inline swarm.go]: here

File: swarm.go [hunk 2]
```diff
@@ -245,7 +245,8 @@ func NewSwarm(config *api.Config, mockStore *mock.NodeStore) (self *Swarm, err e
 	feedsHandler.SetStore(self.netStore)
 
 	syncing := true
-	if !config.SyncEnabled || config.LightNodeEnabled || config.BootnodeMode {
+	// should config.PushSyncEnabled have an effect here too?
+	if !config.PullSyncEnabled || config.LightNodeEnabled || config.BootnodeMode {
```
Comment (mortelli) [inline swarm.go:248]: here

File: swarm.go [hunk 3]
```diff
@@ -562,7 +563,7 @@ func (s *Swarm) APIs() []rpc.API {
 
 	// this is a workaround disabling syncing altogether from a node but
 	// must be changed when multiple stream implementations are at hand
-	if s.config.SyncEnabled {
+	if s.config.PullSyncEnabled { // should config.PushSyncEnabled have an effect here too?
```
Comment (mortelli) [inline swarm.go]: and here

**Review by mortelli — COMMENTED** [review]
i found some places where the—under this implementation deprecated—`no-sync` flag/config is being used to set other values.

now that this flag is split into 2, i am not sure whether both values should be used with the same purpose for these cases, or if we should just check for the pull sync flag.

File: swarm.go (hunk 1 above)
Comment (zelig) [reply swarm.go]: if you replicate the old behaviour you will have pull && push here

File: swarm.go (hunk 2 above)
Comment (zelig) [reply swarm.go:248]: i think  this is ok

**Review by zelig — APPROVED** [review]
approved but marcelo's questions. need to be double checked with @acud

**Review by santicomp2014 — APPROVED** [review]
LGTM

The only part I was not sure was `SyncEnabled:  config.PullSyncEnabled, // should config.PushSyncEnabled have an effect here as well?`

But @zelig cleared my doubt with his comments.

Anyways I think @acud should have a look before merging if the behaviour is correct.

**Review by janos — APPROVED** [review]
Looks OK to me. You can have a look at the PR that removed the --sync-mode flag which handles similar behaviour https://github.com/ethersphere/swarm/pull/1945/files.

**Review by acud — COMMENTED** [review]
@mortelli why do you need to merge this change to `master`? if I understand correctly, this is for manual testing. Why not have this as a single commit which you can revert on the branch? This is not needed for production code, but just for testing. The current flag configuration on `master` IMO is satisfactory and I'm not sure why we need to overload such changes due to manual testing needs.

**Comment by mortelli** [comment]
> @mortelli why do you need to merge this change to `master`? if I understand correctly, this is for manual testing. Why not have this as a single commit which you can revert on the branch? This is not needed for production code, but just for testing. The current flag configuration on `master` IMO is satisfactory and I'm not sure why we need to overload such changes due to manual testing needs.

i would argue that an end-user could make use of these flags, and in fact, 2 separate flags for 2 separate settings seem to make more sense to me. after all, as "end-users" in our manual tests, it was not clear to us that the `no-sync` flag affected push sync as well.

having said that, i'm changing the base branch to `global-pinning` to prevent potential problems with a release if this is a serious concern.

File: swarm.go (hunk 1 above)
Comment (mortelli) [reply swarm.go]: thought this might be the case, thanks

**Review by acud — APPROVED** [review]
:ship:

**Pull request merged**